src/sinks/aws_cloudwatch_logs.rs

```diff
@@ -227,47 +333,93 @@ impl Service<Vec<Event>> for CloudwatchLogsSvc {
     }
 }
 
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+struct CloudwatchKey {
+    group: Bytes,
+    stream: Bytes,
+}
+
+fn interpolate(s: &str) -> Partition {
+    use regex::Regex;
+    let r = Regex::new(r"\{event\.(?P<key>\D+)\}").unwrap();
+
+    if let Some(cap) = r.captures(s) {
+        if let Some(m) = cap.name("key") {
+            return Partition::Event(m.as_str().into());
+        }
+    }
+
+    Partition::Static(s.into())
+}
+
+fn partition(
+    event: Event,
+    group: &Bytes,
+    stream: &Partition,
+) -> impl futures::Stream<Item = PartitionInnerBuffer<Event, CloudwatchKey>, Error = ()> {
```

Comment:
Not sure we need to return a `Stream` here. It looks like an `Option` would work just as well and be much simpler.

I'm also confused why we're returning a whole `PartitionInnerBuffer` with a single event in it. I'd expect us to return a `(Event, CloudwatchKey)` or something like that and those would then get accumulated into a `PartitionInnerBuffer` (still getting over that name) in the next stage.

Reply:
The reason we need Stream here is that we want to use `Sink::with_flat_map` instead of just `Sink::with`. This allows us to drop events if the field we are looking for within the event doesn't exist. At this point, we need to drop the item since we can't possibly make a `stream_name` from this. The stream here is what `Sink::with_flat_map` takes.